File: apps/server/src/services/analysis-jobs.ts
import type {
  NormalizedProduct,
  ProductAnalysisResult,
  AnalysisJobResponse,
  ProductFacts,
  NutritionFacts,
  IngredientAnalysis,
} from '@acme/shared';
import { DEFAULT_ONBOARDING_RESPONSE } from '@acme/shared';
import { randomUUID } from 'node:crypto';

import { getProductFactsService } from './product-facts-ai';
import { searchNutritionData } from './nutrition-websearch';
import { analyzeIngredients } from './ingredient-analysis-ai';
import {
  buildClassificationFromData,
  buildProductFacts,
  buildNutritionFacts,
  hasNutritionData,
} from '../domain/product-facts/build-product-facts';
import { computeAllProfileScores, type ScoreProfileInput } from '../domain/score-engine/compute-score';
import { getProfileInputs } from './profileInputs';
import { updateScanAnalysisResult } from '../repositories/scanRepository';

const JOB_TTL_MS = 10 * 60 * 1000;
const jobs = new Map<string, AnalysisJobResponse>();

const scheduleCleanup = (jobId: string): void => {
  setTimeout(() => {
    jobs.delete(jobId);
  }, JOB_TTL_MS);
};

/**
 * Build ScoreProfileInput[] from userId.
 */
const buildProfiles = async (userId?: string): Promise<ScoreProfileInput[]> => {
  if (!userId) {
    return [{
      profileId: 'you',
      profileType: 'self',
      name: 'You',
      onboarding: DEFAULT_ONBOARDING_RESPONSE,
    }];
  }

  const inputs = await getProfileInputs(userId);
  if (inputs.length === 0) {
    return [{
      profileId: 'you',
      profileType: 'self',
      name: 'You',
      onboarding: DEFAULT_ONBOARDING_RESPONSE,
    }];
  }

  return inputs.map((input, index) => ({
    profileId: input.profileId,
    profileType: index === 0 ? 'self' as const : 'family_member' as const,
    name: input.profileName,
    onboarding: input.onboarding,
  }));
};

/**
 * Run the full analysis pipeline:
 *   1. Get nutrition data from product (OFF/DB). If missing → web search.
 *   2. AI extracts classification facts (productType, diet, nutriGrade) — in parallel with step 1 if web search needed.
 *   3. Merge classification + nutrition → ProductFacts
 *   4. Deterministic score engine computes per-profile scores
 */
const runAnalysisJob = async (
  jobId: string,
  product: NormalizedProduct,
  userId?: string,
  scanId?: string,
): Promise<void> => {
  const productName = product.product_name ?? product.code ?? 'unknown';
  console.log(`\n[Job:${jobId}] ▶ START  product="${productName}"  userId=${userId ?? 'anon'}`);
  const jobStart = Date.now();

  try {
    // Step 1: Build profiles
    const profiles = await buildProfiles(userId);
    const profileNames = profiles.map((p) => `"${p.name}"`).join(', ');
    console.log(`[Job:${jobId}] 👤 Profiles (${profiles.length}): ${profileNames}`);

    // Step 2: Get nutrition facts (product data first, web search fallback)
    const productHasNutrition = hasNutritionData(product);
    console.log(`[Job:${jobId}] 📊 Product has nutrition data: ${productHasNutrition}`);

    // Step 3: AI classification + nutrition resolution + ingredient analysis (all parallel)
    console.log(`[Job:${jobId}] 🧪 Step 1 — Extracting classification${productHasNutrition ? '' : ' + searching nutrition'} + ingredients...`);
    const factsStart = Date.now();

    let classification;
    let nutritionFacts: NutritionFacts;
    let ingredientResults: (IngredientAnalysis | null)[];

    // Ingredient analysis per profile — all run in parallel
    const ingredientPromises = profiles.map((profile) =>
      analyzeIngredients(product, profile.onboarding).catch((err) => {
        console.warn(`[Job:${jobId}] ⚠️ Ingredient analysis failed for "${profile.name}":`, err);
        return null;
      }),
    );

    if (productHasNutrition) {
      // Product has nutrition → just run AI classification + ingredients in parallel
      nutritionFacts = buildNutritionFacts(product);
      const [classificationResult, ...ingResults] = await Promise.all([
        getProductFactsService().extractClassification(product).catch(() => {
          console.warn(`[Job:${jobId}] ⚠️ AI classification failed, using deterministic fallback`);
          return buildClassificationFromData(product);
        }),
        ...ingredientPromises,
      ]);
      classification = classificationResult;
      ingredientResults = ingResults as (IngredientAnalysis | null)[];
    } else {
      // No nutrition → run AI classification + nutrition web search + ingredients in parallel
      const [classificationResult, webNutrition, ...ingResults] = await Promise.all([
        getProductFactsService().extractClassification(product).catch(() => {
          console.warn(`[Job:${jobId}] ⚠️ AI classification failed, using deterministic fallback`);
          return buildClassificationFromData(product);
        }),
        searchNutritionData(productName, product.brands, product.code),
        ...ingredientPromises,
      ]);

      classification = classificationResult;
      nutritionFacts = webNutrition ?? buildNutritionFacts(product);
      ingredientResults = ingResults as (IngredientAnalysis | null)[];

      if (webNutrition) {
        console.log(`[Job:${jobId}] 🌐 Nutrition found via web search`);
      } else {
        console.log(`[Job:${jobId}] ⚠️ No nutrition found anywhere — scoring with empty data`);
      }
    }

    // Build per-profile ingredient analysis map
    const perProfileIngredients = new Map<string, IngredientAnalysis | null>();
    for (let i = 0; i < profiles.length; i++) {
      perProfileIngredients.set(profiles[i].profileId, ingredientResults[i] ?? null);
    }

    // Step 4: Merge classification + nutrition → ProductFacts
    const facts: ProductFacts = buildProductFacts(classification, nutritionFacts);
    console.log(`[Job:${jobId}] ✅ Facts built  ${Date.now() - factsStart}ms  type=${facts.productType}`);

    // Step 5: Deterministic score engine (per-profile ingredient analysis)
    console.log(`[Job:${jobId}] 🧮 Step 2 — Computing scores...`);
    const scoreStart = Date.now();
    const profileScores = computeAllProfileScores(facts, profiles, perProfileIngredients);
    console.log(`[Job:${jobId}] ✅ Scores computed  ${Date.now() - scoreStart}ms`);

    for (const ps of profileScores) {
      console.log(`[Job:${jobId}]   "${ps.name}" → score=${ps.score} (${ps.fitLabel})  +${ps.positives.length}✓ -${ps.negatives.length}✗`);
    }

    // Step 6: Build final result
    const selfIngredientAnalysis = perProfileIngredients.get(
      profiles.find((p) => p.profileType === 'self')?.profileId ?? 'you',
    );

    const result: ProductAnalysisResult = {
      productFacts: facts,
      profiles: profileScores,
      ...(selfIngredientAnalysis ? { ingredientAnalysis: selfIngredientAnalysis } : {}),
    };

    const job = jobs.get(jobId);
    if (job) {
      job.status = 'completed';
      job.result = result;
    }

    // Step 5: Persist to database
    if (scanId) {
      await updateScanAnalysisResult(scanId, 'completed', result).catch(() => {});
    }

    console.log(`[Job:${jobId}] 🏁 Job completed  total=${Date.now() - jobStart}ms`);
  } catch (err) {
    console.error(`[Job:${jobId}] ❌ Job failed:`, err);
    const job = jobs.get(jobId);
    if (job) {
      job.status = 'failed';
    }
    if (scanId) {
      await updateScanAnalysisResult(scanId, 'failed').catch(() => {});
    }
  }
};

/**
 * Create a new analysis job. Returns immediately with a pending job reference.
 */
export const createAnalysisJob = (
  product: NormalizedProduct,
  userId?: string,
  scanId?: string,
): { jobId: string; status: 'pending' } => {
  const jobId = randomUUID();
  const job: AnalysisJobResponse = {
    jobId,
    status: 'pending',
  };

  jobs.set(jobId, job);
  scheduleCleanup(jobId);
  void runAnalysisJob(jobId, product, userId, scanId);

  return { jobId, status: 'pending' };
};

/**
 * Create a pre-completed job from cached results.
 */
export const createCachedAnalysisJob = (
  cachedResult: ProductAnalysisResult,
): { jobId: string; status: 'completed' } => {
  const jobId = randomUUID();

  jobs.set(jobId, {
    jobId,
    status: 'completed',
    result: cachedResult,
  });
  scheduleCleanup(jobId);

  return { jobId, status: 'completed' };
};

/**
 * Get a job by its ID.
 */
export const getAnalysisJob = (jobId: string): AnalysisJobResponse | null => {
  return jobs.get(jobId) ?? null;
};

File: apps/server-py/app/services/analysis_jobs.py
"""In-memory async analysis job manager.

Each job runs the full product analysis pipeline in the background:
  1. Build user profiles from DB
  2. AI classification (productType, dietCompatibility, nutriGrade)
  3. Nutrition websearch if missing
  4. Ingredient analysis per profile (parallel)
  5. Score engine → per-profile scores
"""

import asyncio
from typing import Optional
from uuid import uuid4

from loguru import logger

from app.domain.product_facts.build_product_facts import (
    build_classification_from_data,
    build_product_facts,
    has_nutrition_data,
)
from app.domain.product_facts.nutrition_utils import build_nutrition_facts
from app.domain.product_facts.schema import (
    IngredientAnalysis,
    NormalizedProduct,
    OnboardingProfile,
    ProductAnalysisResult,
)
from app.domain.score_engine.compute_score import compute_all_profile_scores
from app.services.ingredient_analysis_ai import analyze_ingredients
from app.services.nutrition_websearch import search_nutrition_data
from app.services.product_facts_ai import get_product_facts_service

_JOB_TTL_S = 10 * 60

_jobs: dict[str, dict] = {}


def _default_profile() -> OnboardingProfile:
    return {
        "profileId": "you",
        "profileType": "self",
        "name": "You",
        "mainGoal": None,
        "restrictions": [],
        "allergies": [],
        "otherAllergiesText": None,
        "nutritionPriorities": [],
        "onboardingCompleted": False,
    }


async def _build_profiles(user_id: Optional[str]) -> list[OnboardingProfile]:
    """Load user + family profiles from DB. Falls back to a single anonymous profile."""
    if not user_id:
        return [_default_profile()]

    try:
        from app.utils.unitofwork import UnitOfWork

        async with UnitOfWork() as uow:
            user = await uow.users.get_one_or_none(id=user_id)
            if user is None:
                return [_default_profile()]

            profile = await uow.user_profiles.get_one_or_none(user_id=user_id)
            family_members = await uow.family_members.get_multi(offset=0, limit=50, user_id=user_id)

        profiles: list[OnboardingProfile] = []

        # Self profile
        if profile:
            profiles.append(
                {
                    "profileId": user_id,
                    "profileType": "self",
                    "name": user.name or "You",
                    "mainGoal": profile.main_goal,
                    "restrictions": list(profile.restrictions or []),
                    "allergies": list(profile.allergies or []),
                    "otherAllergiesText": profile.other_allergies_text,
                    "nutritionPriorities": list(profile.nutrition_priorities or []),
                    "onboardingCompleted": profile.onboarding_completed,
                }
            )
        else:
            self_profile = _default_profile()
            self_profile["name"] = user.name or "You"
            self_profile["profileId"] = user_id
            profiles.append(self_profile)

        # Family member profiles
        for member in family_members:
            profiles.append(
                {
                    "profileId": str(member.id),
                    "profileType": "family_member",
                    "name": member.name or "Family member",
                    "mainGoal": getattr(member, "main_goal", None),
                    "restrictions": list(getattr(member, "restrictions", None) or []),
                    "allergies": list(getattr(member, "allergies", None) or []),
                    "otherAllergiesText": getattr(member, "other_allergies_text", None),
                    "nutritionPriorities": list(getattr(member, "nutrition_priorities", None) or []),
                    "onboardingCompleted": True,
                }
            )

        return profiles if profiles else [_default_profile()]

    except Exception as exc:
        logger.warning(f"[Jobs] Failed to load profiles for {user_id}: {exc}")
        return [_default_profile()]


async def _run_job(
    job_id: str,
    product: NormalizedProduct,
    user_id: Optional[str] = None,
    scan_id: Optional[str] = None,
) -> None:
    product_name = product.get("product_name") or product.get("code") or "unknown"
    logger.info(f"[Job:{job_id}] ▶ START product='{product_name}' user={user_id or 'anon'}")

    try:
        # Step 1: Build profiles
        profiles = await _build_profiles(user_id)
        logger.debug(f"[Job:{job_id}] profiles: {[p['name'] for p in profiles]}")

        product_has_nutrition = has_nutrition_data(product)
        logger.debug(f"[Job:{job_id}] has nutrition: {product_has_nutrition}")

        # Step 2: AI classification + nutrition search + ingredient analysis (all parallel)
        ingredient_promises = [
            asyncio.create_task(_safe_analyze_ingredients(product, profile, job_id)) for profile in profiles
        ]

        if product_has_nutrition:
            nutrition_facts = build_nutrition_facts(product)
            classification_result, *ing_results = await asyncio.gather(
                _safe_extract_classification(product, job_id),
                *ingredient_promises,
            )
        else:
            classification_result, web_nutrition, *ing_results = await asyncio.gather(
                _safe_extract_classification(product, job_id),
                search_nutrition_data(
                    product_name,
                    product.get("brands"),
                    product.get("code"),
                ),
                *ingredient_promises,
            )
            nutrition_facts = web_nutrition or build_nutrition_facts(product)
            if web_nutrition:
                logger.info(f"[Job:{job_id}] nutrition found via web search")
            else:
                logger.warning(f"[Job:{job_id}] no nutrition found anywhere — scoring with empty data")

        per_profile_ingredients: dict[str, Optional[IngredientAnalysis]] = {
            profiles[i]["profileId"]: ing_results[i] for i in range(len(profiles))
        }

        # Step 3: Build ProductFacts
        facts = build_product_facts(classification_result, nutrition_facts)

        # Step 4: Compute scores
        profile_scores = compute_all_profile_scores(facts, profiles, per_profile_ingredients)

        for ps in profile_scores:
            logger.info(
                f"[Job:{job_id}] '{ps['name']}' → score={ps['score']} ({ps['fitLabel']}) "
                f"+{len(ps['positives'])}✓ -{len(ps['negatives'])}✗"
            )

        # Step 5: Build result
        self_profile = next((p for p in profiles if p["profileType"] == "self"), None)
        self_ingredients = per_profile_ingredients.get(self_profile["profileId"]) if self_profile else None

        result: ProductAnalysisResult = {
            "productFacts": facts,
            "profiles": profile_scores,
        }
        if self_ingredients:
            result["ingredientAnalysis"] = self_ingredients

        # Update job
        job = _jobs.get(job_id)
        if job:
            job["status"] = "completed"
            job["result"] = result

        # Persist to DB
        if scan_id:
            await _persist_scan_result(scan_id, "completed", result)

        logger.info(f"[Job:{job_id}] ✓ done")

    except Exception as exc:
        logger.error(f"[Job:{job_id}] failed: {exc}")
        job = _jobs.get(job_id)
        if job:
            job["status"] = "failed"
        if scan_id:
            await _persist_scan_result(scan_id, "failed", None)


async def _safe_extract_classification(product: NormalizedProduct, job_id: str):
    try:
        return await get_product_facts_service().extract_classification(product)
    except Exception as exc:
        logger.warning(f"[Job:{job_id}] AI classification failed: {exc}")
        return build_classification_from_data(product)


async def _safe_analyze_ingredients(
    product: NormalizedProduct, profile: OnboardingProfile, job_id: str
) -> Optional[IngredientAnalysis]:
    try:
        return await analyze_ingredients(product, profile)
    except Exception as exc:
        logger.warning(f"[Job:{job_id}] ingredient analysis failed for '{profile['name']}': {exc}")
        return None


async def _persist_scan_result(scan_id: str, status: str, result) -> None:
    try:
        from app.utils.unitofwork import UnitOfWork

        update_data = {
            "personal_analysis_status": status,
            "personal_analysis_job_id": None,
        }
        if result:
            update_data["multi_profile_result"] = result

        async with UnitOfWork() as uow:
            await uow.scans.update(update_data, return_object=False, id=scan_id)

    except Exception as exc:
        logger.warning(f"[Jobs] Failed to persist scan result {scan_id}: {exc}")


def _schedule_cleanup(job_id: str) -> None:
    async def _cleanup():
        await asyncio.sleep(_JOB_TTL_S)
        _jobs.pop(job_id, None)

    asyncio.create_task(_cleanup())


def create_analysis_job(
    product: NormalizedProduct,
    user_id: Optional[str] = None,
    scan_id: Optional[str] = None,
) -> dict:
    job_id = str(uuid4())
    job = {"jobId": job_id, "status": "pending"}
    _jobs[job_id] = job
    _schedule_cleanup(job_id)
    asyncio.create_task(_run_job(job_id, product, user_id, scan_id))
    return {"jobId": job_id, "status": "pending"}


def create_cached_analysis_job(cached_result: dict) -> dict:
    job_id = str(uuid4())
    _jobs[job_id] = {"jobId": job_id, "status": "completed", "result": cached_result}
    _schedule_cleanup(job_id)
    return {"jobId": job_id, "status": "completed"}


def get_analysis_job(job_id: str) -> Optional[dict]:
    return _jobs.get(job_id)
